### ISSUE TITLE
False positive for global-require

### ISSUE THREAD

**Issue by FM-96** [eslint/eslint#8784]
**Tell us about your environment**

* **ESLint Version:** 4.0.0
* **Node Version:** 7.10.0
* **npm Version:** 5.0.3

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**

```json
{
	"env": {
		"es6": true,
		"node": true
	},
	"extends": "eslint:recommended",
	"rules": {
		"global-require": "error"
	}
}
```

**What did you do? Please include the actual source code causing the issue.**
This code is at the top level of the file:

```js
var botVersion = process.env.npm_package_version || require('./package.json').version;
```

**What did you expect to happen?**
I expect this code to work without warnings, as the statement is at the top level of the file.
Notably, as the documentation of the global-require rule states, using a ternary operator instead gives no warnings:
```js
var botVersion = process.env.npm_package_version ? process.env.npm_package_version : require('./package.json').version;
```
And the documentation for the no-unneeded-ternary rule asserts that using a logical OR operator should provide "the same functionality" as a ternary in this case.

**What actually happened? Please include the actual, raw output from ESLint.**
ESLint gives a warning:
```
1:53   error    Unexpected require()                                            global-require
```

**Comment by mysticatea** [comment]
Thank you for the report.

However, this works as intended.
As https://github.com/eslint/eslint/issues/5773#issuecomment-205868680, `global-require` rule reports `require()` expressions which cannot replace by ES2015 Modules. So the rule disallows conditional importing.

**Comment by platinumazure** [comment]
It's a bit inconsistent that this rule allows ternary conditional importing but doesn't allow logical expression conditional importing. Are you sure there's nothing we can do here @mysticatea?

**Comment by FM-96** [comment]
@mysticatea 
But... that's incorrect. The rule *doesn't* disallow conditional importing. Using a ternary expression to determine which module to require is explicitly allowed as per the [documentation](http://eslint.org/docs/rules/global-require).
Which means if the goal of the rule is to emulate `import` expressions (which by the way is not evident at all from the documentation) then it fails at that too, as those afaik cannot be used in ternary expressions.

So either way, something is not working correctly here.

**Comment by ilyavolodin** [comment]
I agree. Either this rule shouldn't be checking logical expressions, or it should also check ternaries.

**Comment by mysticatea** [comment]
Oh, I'm sorry for overlooking.
Hmm, I'd like to hear author's opinion.

**Comment by FM-96** [comment]
So... is there anything else happening regarding this issue?

**Comment by ljharb** [comment]
imo the bug is that it doesn't also check ternaries.

If people want to allow module-level dynamic `require`s, I think that should be a separate rule (and this one disabled), or an option to this rule.

**Comment by not-an-aardvark** [comment]
From https://github.com/eslint/eslint/issues/8942, the rule also reports

```js
var a = require('./b') - 1;
```

**Comment by not-an-aardvark** [comment]
I think the problem here is that the [rule documentation](http://eslint.org/docs/rules/global-require) describes the "top level of a module", but that doesn't have a well-defined meaning. It can either mean "not inside any functions", or it can mean "in a position where it could be replaced with an `import` statement".

Either of these would be relatively simple to implement, but until we decide what the rule is trying to do in the first place, I don't we can decide whether any behavior in between is a bug. (As a separate note, given that using ternaries is explicitly allowed in the docs, I'm not sure it would be reasonable for us to start disallowing them by default and call it a bugfix.)

**Comment by ljharb** [comment]
It has a very well-defined meaning - it means in the top-level scope; which means "not inside any curly braces".

**Comment by not-an-aardvark** [comment]
Do you mean that the following should be considered in the top-level scope? Neither of them are inside any curly braces.

```js
const bar = x ? require('baz') : require('qux');
const foo = () => require('a');
```

I'm not trying to be nitpicky -- I just want to fully understand the behavior you're describing. Describing it in terms of lexical variable scopes makes sense and would be consistent, but it would result in allowing ternaries, which I thought you were opposed to from https://github.com/eslint/eslint/issues/8784#issuecomment-313285770.

**Comment by ljharb** [comment]
Lol, ok, curly braces or implicit blocks, like if/else/do/while and implicit-return arrow functions.

I think this rule should be erroring on any non-static non-top-level non-unconditional require, which means ternaries - while at the top level, and in this case static - would still error based on being conditional.

**Comment by ljharb** [comment]
To clarify, I think that this rule deals with enforcing that requires are like import statements, not that they're "top-level".

**Comment by not-an-aardvark** [comment]
Should this be allowed?

```js
const foo = require('bar').baz();
```

`global-require` currently allows it, but it wouldn't be possible to convert it to an import statement because it would require two statements if `import` was used (one for the import, and one for the method call).

**Comment by ljharb** [comment]
Yes, it should be allowed - I don't think the rule should be preventing "requires that can't be come an import in one statement" - only requires that can't become an import.

**Comment by not-an-aardvark** [comment]
I see, thanks for explaining. What you're describing seems like it would be useful to enforce, but I'm concerned that it would be a significant change from the current behavior of the `global-require` rule (the current rule doesn't care whether a `require` is dynamic, and it explicitly allows conditional requires).

**Comment by ljharb** [comment]
In that case, the original example should pass.

**Comment by not-an-aardvark** [comment]
So if we allow conditional requires and dynamic requires (since the previous version of the rule allowed them), would the rule just disallow requires outside the top-level lexical scope?

**Comment by platinumazure** [comment]
It seems to me that this rule might be taking on too much responsibility. I think we should consider splitting this into two rules with ESLint 5.x:

* global-require ensures that any call to (non-shadowed) `require()` must be in global or module scope, i.e., not in a block or function scope. It does not care about how the `require()` is invoked.
* New rule no-conditional-require ensures that `require()` calls cannot be in a ConditionalExpression or LogicalExpression.

Would that make sense?

**Comment by not-an-aardvark** [comment]
All of that behavior can be enforced with `no-restricted-syntax` anyway, so I'm a bit averse to adding new rules for it.

**Comment by ljharb** [comment]
Lumping everything into `no-restricted-syntax` makes it really hard to configure in a shared config, because there's no way to specify in JSON/YML that I want to *add* to the existing array of restricted syntaxes - which means I have to gather up and copy/paste the entire list from wherever it's defined.

**Comment by ljharb** [comment]
Also, `require` isn't syntax.

**Comment by not-an-aardvark** [comment]
> Lumping everything into `no-restricted-syntax` makes it really hard to configure in a shared config, because there's no way to specify in JSON/YML that I want to *add* to the existing array of restricted syntaxes - which means I have to gather up and copy/paste the entire list from wherever it's defined.

I agree. I've been meaning to make a proposal to make that easier.

> Also, `require` isn't syntax.

You can select it with `CallExpression[callee.name="require"]`.

**Comment by ljharb** [comment]
Still, I don't think the existence of AST selectors and generic rules should be a reason to not make specialized, discoverable, easy-to-use rules. In this case, legislating how `require` is used is a pretty common use case.

I'd say perhaps a single rule that governs all aspects of `node`'s `require`, with options to control the various things, and that bails out when the env isn't `node`, would be useful.

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
